src/app/qgisapp.cpp

```diff
@@ -9588,6 +9589,47 @@ void QgisApp::pasteFromClipboard( QgsMapLayer *destinationLayer )
   // now create new feature using pasted feature as a template. This automatically handles default
   // values and field constraints
   QgsFeatureList newFeatures {QgsVectorLayerUtils::createFeatures( pasteVectorLayer, newFeaturesDataList, &context )};
+
+  // check constraints
+  QgsFeatureList validFeatures = newFeatures;
+  QgsFeatureList invalidFeatures;
+  for ( const QgsFeature &f : qgis::as_const( validFeatures ) )
+  {
+    for ( int idx = 0; idx < pasteVectorLayer->fields().count(); ++idx )
+    {
+      QStringList errors;
+      if ( !QgsVectorLayerUtils::validateAttribute( pasteVectorLayer, f, idx, errors, QgsFieldConstraints::ConstraintStrengthHard, QgsFieldConstraints::ConstraintOriginNotSet ) )
+      {
+        invalidFeatures << f;
+        validFeatures.removeOne( f );
```

Comment:
You can't modify a qlist while iterating over it like this. You'd need to use a modifiable iterator instead and .erase the current entry.